// selenium-appium/dist/index.js
/**
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License.
 */
'use strict';
Object.defineProperty(exports, "__esModule", { value: true });
exports.Config = exports.windowsAppDriverCapabilities = exports.PageObject = exports.WebDriver2 = exports.driver = exports.By2 = void 0;
var by2_1 = require("./by2");
Object.defineProperty(exports, "By2", { enumerable: true, get: function () { return by2_1.By2; } });
var Driver_1 = require("./Driver");
Object.defineProperty(exports, "driver", { enumerable: true, get: function () { return Driver_1.driver; } });
Object.defineProperty(exports, "WebDriver2", { enumerable: true, get: function () { return Driver_1.WebDriver2; } });
var pageobject_1 = require("./pageobject");
Object.defineProperty(exports, "PageObject", { enumerable: true, get: function () { return pageobject_1.PageObject; } });
var WinAppDriver_1 = require("./WinAppDriver");
Object.defineProperty(exports, "windowsAppDriverCapabilities", { enumerable: true, get: function () { return WinAppDriver_1.windowsAppDriverCapabilities; } });
var Config_1 = require("./Config");
Object.defineProperty(exports, "Config", { enumerable: true, get: function () { return Config_1.Config; } });
